non-critical-infra/modules/matrix-synapse.nix

```diff
@@ -0,0 +1,130 @@
+{ config
+, pkgs
+, ...
+}:
+
+{
+  imports = [
+    ./nginx.nix
+    ./postgresql.nix
+  ];
+
+  fileSystems."/var/lib/matrix-synapse" = {
+    device = "zroot/root/matrix-synapse";
+    fsType = "zfs";
+    options = [
+      "zfsutil"
+    ];
+  };
+
+  services.postgresql = {
+    ensureUsers = [ {
+      name = "matrix-synapse";
+      ensureDBOwnership = true;
+    } ];
+    ensureDatabases = [
+      "matrix-synapse"
+    ];
+  };
+
+  services.redis.servers.matrix-synapse = {
+    enable = true;
+  };
+
+  environment.systemPackages = with pkgs; [
+    matrix-synapse-tools.synadm
+  ];
+
+  services.backup = {
+    preHook = ''
+      ${config.boot.zfs.package}/bin/zfs snapshot -r ${config.fileSystems."/var/lib/matrix-synapse".device}@backup
+    '';
+    includes = [
+      "/var/lib/matrix-synapse/.snapshot/backup"
+    ];
+    postHook = ''
+      ${config.boot.zfs.package}/bin/zfs destroy -r ${config.fileSystems."/var/lib/matrix-synapse".device}@backup
+    '';
+  };
+
+  # create /run/matrix-synapse for the unix socket listener
+  systemd.services.matrix-synapse.serviceConfig.RuntimeDirectory = "matrix-synapse";
+
+  services.matrix-synapse = {
+    enable = true;
+    withJemalloc = true;
+
+    # https://github.com/element-hq/synapse/blob/master/docs/usage/configuration/config_documentation.md
+    settings = {
+      enable_metrics = true;
+
+      server_name = "nixos.org";
+      public_base = "https://matrix.nixos.org";
+      admin_contact = "infra@nixos.org";
+
+      allow_public_rooms_over_federation = true;
+      allow_public_rooms_without_auth = true;
+
+      max_upload_size = "50M";
+
+      database = {
+        name = "psycopg2";
+        args = {
+          host = "/run/postgresql";
+        };
+      };
+
+      redis = {
+        enabled = true;
+        path = config.services.redis.matrix-synapse.path;
+      };
+
+      listeners = [ {
+        type = "http";
+        path = "/run/matrix-synapse/matrix-synapse.sock";
+        mode = "0660";
+        resources = [ {
+          compress = true;
+          names = [
+            "client"
+          ];
+        } {
+          compress = false;
+          names = [
+            "federation"
+          ];
+        } ];
+      } {
+        type = "http";
+        bind_addresses = [
+          "127.0.0.1"
+          "::1"
+        ];
+        port = 8090;
+        resources = [ {
+          names = [
+            "metrics"
+          ];
+        } ];
+      } ];
+    };
+  };
+
+  services.nginx = {
+    clientMaxBodySize = config.services.matrix-synapse.settings.max_upload_size;
+    upstreams."matrix-synapse" = {
+      "unix:/run/matrix-synapse/matrix-synapse.sock" = {};
+    }; 
+    virtualHosts."matrix.nixos.org" = {
+      forceSSL = true;
+      enableACME = true;
+
+      locations."^(/_matrix|/_synapse/client)" = {
```

Comment:
Don't you also need `/.well-known/matrix/server` to delegate the homeserver for `nixos.org` to `matrix.nixos.org`?

Reply:
Yeah, but that will happen over in nixos-homepage, since we're hosting nixos.org.

Reply:
Ah right, wasn't aware that this isn't part of the configs in here, disregard then.

Reply:
https://github.com/NixOS/nixos-homepage/pull/1218